File: internal/notifier/notifier.go
// Package notifier 提供通知功能，当获取上游数据失败时发送通知。
package notifier

import (
	"fmt"
	"log/slog"
	"time"
)

// ServiceConfig 是单个通知服务的配置。
type ServiceConfig struct {
	Type       string // feishu, ...
	WebhookURL string
	Secret     string // 飞书签名密钥（可选）
}

// Config 是通知器的配置。
type Config struct {
	Enabled     bool
	ServiceURLs []string        // 兼容旧配置（仅打日志）
	Services    []ServiceConfig // 通知服务列表
}

// sender 是通知发送接口。
type sender interface {
	Send(message string) error
}

// Notifier 是通知管理器。
type Notifier struct {
	enabled bool
	senders []sender
}

// New 创建一个新的通知器实例。
func New(cfg Config) *Notifier {
	var senders []sender
	for _, svc := range cfg.Services {
		switch svc.Type {
		case "feishu":
			senders = append(senders, newFeishuSender(svc.WebhookURL, svc.Secret))
		default:
			slog.Warn("未知的通知服务类型，已跳过", "type", svc.Type)
		}
	}

	return &Notifier{
		enabled: cfg.Enabled,
		senders: senders,
	}
}

// Notify 发送通知。
func (n *Notifier) Notify(feedKey string, statusCode int, errorMessage string) {
	if !n.enabled || len(n.senders) == 0 {
		return
	}

	message := fmt.Sprintf(
		"[RSSGen] 订阅源 %s 获取失败\n状态码: %d\n错误: %s\n时间: %s",
		feedKey,
		statusCode,
		errorMessage,
		time.Now().Format("2006-01-02 15:04:05"),
	)

	// 异步发送通知，不阻塞主流程
	go n.sendNotification(message)
}

func (n *Notifier) sendNotification(message string) {
	slog.Info("发送通知", "message", message)
	for _, s := range n.senders {
		if err := s.Send(message); err != nil {
			slog.Error("通知发送失败", "error", err)
		}
	}
}

File: RSSGen/core/notifier.py
"""通知模块：当获取上游数据失败时发送通知"""

from datetime import datetime

from loguru import logger


class Notifier:
    """通知管理器"""

    # 默认业务错误状态码
    DEFAULT_BUSINESS_ERROR_CODES = [400, 401, 403, 404, 410, 422, 451]

    def __init__(self, config: dict):
        """初始化 notifier，加载配置"""
        notifier_config = config.get("notifier", {})
        self.enabled = notifier_config.get("enabled", False)
        self.service_urls = notifier_config.get("service_urls", [])
        self.business_error_codes = set(
            notifier_config.get("business_error_codes", self.DEFAULT_BUSINESS_ERROR_CODES)
        )
        self.disabled_feeds: set[str] = set()
        self._apprise = None

    def is_business_error(self, status_code: int) -> bool:
        """判断是否为业务错误"""
        return status_code in self.business_error_codes

    def is_feed_disabled(self, feed_key: str) -> bool:
        """检查 feed 是否被禁用（feed_key 即 cache_key：route/feed_id）"""
        return feed_key in self.disabled_feeds

    def disable_feed(self, feed_key: str):
        """禁用单个 feed（feed_key 即 cache_key：route/feed_id）"""
        self.disabled_feeds.add(feed_key)

    async def notify(self, feed_key: str, status_code: int, error_message: str):
        """发送通知"""
        if not self.enabled or not self.service_urls:
            return

        message = f"[RSSGen] 订阅源 {feed_key} 获取失败\n"
        message += f"状态码: {status_code}\n"
        message += f"错误: {error_message}\n"
        message += f"时间: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}"

        await self._send_notification(message)

    async def _send_notification(self, message: str):
        """实际发送通知的内部方法"""
        try:
            if self._apprise is None:
                import apprise

                self._apprise = apprise.Apprise()
                for url in self.service_urls:
                    self._apprise.add(url)

            # apprise.notify 是同步阻塞方法，放到线程池执行避免阻塞事件循环
            import asyncio

            result = await asyncio.to_thread(self._apprise.notify, message)

            if result:
                logger.info("已发送通知")
            else:
                logger.error("发送通知失败：Apprise 返回 False")
        except Exception as e:
            logger.error(f"发送通知失败: {e}")
